File: Data Structures & Algorithms/permutations/submission-2.java
class Solution {

    public void dfs(int i, List<Integer> curr, List<List<Integer>> result, boolean[] used, int[] nums)
    {
        if(curr.size() == nums.length)
        {
            result.add(new ArrayList<Integer>(curr));
            return;
        }
        for(int j = 0; j < used.length; j++)
        {
            if(used[j] == false)
            {
                curr.add(nums[j]);
                used[j] = true;
                dfs(j, curr, result, used, nums);
                curr.remove(curr.size()-1);
                used[j] = false;
            }
        }
        
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        dfs(0, curr, result, used, nums);
        return result;
    }
}

File: Data Structures & Algorithms/surrounded-regions/submission-2.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        d = [[0,1], [0,-1], [1,0],[-1,0]]
        def dfs(x, y):
            if x >= len(board) or x < 0 or y < 0 or y >= len(board[0]):
                return
            if board[x][y] != "O":
                return
            board[x][y] = "S"
            for dx, dy in d:
                dfs(dx+x, dy+y)

        for x in range(len(board)):
            if board[x][0] == "O":
                dfs(x, 0)
            if board[x][len(board[0])-1] == "O":
                dfs(x, len(board[0])-1)
        for y in range(1, len(board[0])-1):
            if board[0][y] == "O":
                dfs(0, y)
            if board[len(board)-1][y] == "O":
                dfs(len(board)-1, y)
        
        for x in range(len(board)):
            for y in range(len(board[0])):
                if board[x][y] == "O":
                    board[x][y] = "X"
                if board[x][y] == "S":
                    board[x][y] = "O"